pshtt.py

```diff
@@ -101,87 +102,125 @@ def result_for(domain):
-        if endpoint.protocol == "http":
-            endpoint.redirect = True
-    # Endpoint is not live
-    # TODO: Too broad, shouldn't swallow all errors.
-    # except:
-    #     logging.debug("Endpoint is not live: %s" % endpoint.url)
-    #     pass
+        # Retry with certificate validation disabled.
+        try:
+            req = ping(endpoint.url, verify=False)
+        except requests.exceptions.SSLError:
+            # If it's a protocol error or other, it's not live.
+            endpoint.live = False
+            return
+
+        # If it was a certificate error of any kind, it's live.
+        # Figure out the error(s).
+        https_check(endpoint)
+
+    # This needs to go last, as a parent error class.
+    except requests.exceptions.ConnectionError:
+        endpoint.live = False
+        return
 
 
-def https_check(endpoint):
-    logging.debug("sslyzing %s..." % endpoint.url)
+    # Endpoint is live, analyze the response.
+    endpoint.live = True
+    endpoint.headers = dict(req.headers)
 
-    # Use sslyze to check for HSTS and certificate errors
-    try:
-        # remove the https:// from prefix for sslyze
-        hostname = endpoint.url[8:]
-        server_info = ServerConnectivityInfo(hostname=hostname, port=443)
-        server_info.test_connectivity_to_server()
-
-        # Call plugin directly
-        cert_plugin = CertificateInfoPlugin()
-        cert_plugin_result = cert_plugin.process_task(server_info, 'certinfo_basic')
-        # Parsing sslyze output for results by line
-        for i in cert_plugin_result.as_text():
-            # Check for cert expiration
-            if "Not After" in i:
-                expired_cert(i, endpoint)
-            # Check for Hostname validation
-            elif "Hostname Validation" in i:
-                bad_hostname(i, endpoint)
-            # Check if Cert is trusted based on CA Stores
-            elif "CA Store" in i:
-                bad_chain(i, endpoint)
-                break
-    except:
-        # No valid hsts
-        pass
+    endpoint.status = req.status_code
+    if str(endpoint.status).startswith('3'):
+        endpoint.redirect = True
+
+    if endpoint.redirect:
+
+        location_header = req.headers.get('Location')
+        # Absolute redirects (e.g. "https://example.com/Index.aspx")
+        if location_header.startswith("http:") or location_header.startswith("https:"):
+            immediate = location_header
+
+        # Relative redirects (e.g. "Location: /Index.aspx").
+        # Construct absolute URI, relative to original request.
+        else:
+            immediate = urlparse.urljoin(endpoint.url, location_header)
+
+        # Chase down the ultimate destination, ignoring any certificate warnings.
+        # TODO: try/except block on this request, and have checks expect possible None.
+        ultimate_req = ping(endpoint.url, allow_redirects=True, verify=False)
```

Comment:
This probably needs a `try`/`except` block around it too to handle connection errors when following redirects, even though we know that the endpoint is live at this point.